feat: implement mint command and rename binaries

Keinberger opened the pull request:
# Summary

This PR implements a `mint` CLI command, that users can use to request a public P2ID note from a faucet, i.e. "mint" tokens from a faucet.

Furthermore, this PR introduces a new `miden-faucet-client` binary, which implements the "mint" functionality. The reasoning as to why this decision was made is described in more detail at the "Binary refactoring" section below. This explains the large amount of file changes, although most are just due to renaming the existing "faucet" binary folder.

The `mint` command is part of the `miden mint` command that will be implemented at `midenup`. See this comment here on how the `miden mint` command will work on `midenup`: https://github.com/0xMiden/miden-faucet/issues/192#issuecomment-3625918936

This feature was discussed on this issue: https://github.com/0xMiden/miden-faucet/issues/192

## Usage

The new command can be used as follows:

```bash
miden-faucet-client mint --amount <AMOUNT> --account <ACCOUNT_ID> --url <FAUCET_URL>
```

Note that the `--url` parameter is optional, the faucet will default to the testnet faucet URL if the flag has not been provided.

## Open Questions

- There is a helper function that I implemented here: I borrowed this helper function from the code present at the `miden-client` repository. Since both `miden-client` and `miden-faucet` will now use the same logic, I think it might make sense to include that function in `miden-base` so that both repositories can import it from there. Let me know about your thoughts!

- I placed the tests inside of the `miden-faucet-client` into a dedicated `tests/` folder, which speaks against the fashion of how tests are laid out across the rest of the faucet repository. I think it's still better in terms of code organization to have them in a dedicated `tests/´ folder, but happy to move them inside of the `mint.rs` file directly if that's what we prefer.

Closes #192 

---

## EDITS

### CI `action.yml` file changes

As part of this PR, I made changes to the `action.yml` file workflow to make sure that it references all crate, file and directory names correctly (since some have been renamed). Checking these changes for accuracy is still pending. I have outlined those changes in the following comment: https://github.com/0xMiden/miden-faucet/pull/196#issuecomment-3669669992

### Request Struct exports

As outlined in the following comment, the PR also changes the `faucet-operator` to export request structs that are used by the `faucet-client` for making requests to a faucet: https://github.com/0xMiden/miden-faucet/pull/196#issuecomment-3665135737

## Binary refactoring

The new `mint` command is implemented inside of a dedicated binary called `miden-faucet-client`.

The two libraries will and are responsible for holding distinct logic:

- `miden-faucet`: Existing `miden-faucet` binary.
- `miden-faucet-client`: This one contains the mint logic (and any future logic we may want to add to interact with live faucets). Not sure if the naming is perfect, let me know if you have any other proposals!

This has been outlined in the following comment: https://github.com/0xMiden/miden-faucet/pull/196#issuecomment-3642851097

Throughout the development of this PR, the faucet binary was temporarily renamed to `miden-faucet-operator`. However, after discussion with reviewers, we've reverted this change to maintain backward compatibility. The final implementation keeps the original `miden-faucet` binary name unchanged, while adding the new `miden-faucet-client` binary. This avoids unnecessary infrastructure changes and deployment work. See comment: https://github.com/0xMiden/miden-faucet/pull/196#issuecomment-3759672160

lima-limon-inc reviewed (commented):
Overall looks good, I tested on my machine a couple of times* and managed to mint tokens, here's the [midenscan link](https://testnet.midenscan.com/tx/0x93bd0afd26def4cda11b932caf77e727c65c623933803d2f4607a57ab2adb6e0).

I think my main question is if whether encapsulate the functionality in a separate library and also a separate binary. Would love to hear other thoughts!

* save for this [small issue](https://github.com/0xMiden/miden-faucet/pull/196#discussion_r2607128985)

Keinberger commented on bin/faucet/src/main.rs:
```diff
@@ -2,6 +2,7 @@ mod api;
 mod api_key;
 mod frontend;
 mod logging;
+mod mint;
```
Yeah I think that could make sense. How would you go about naming that library, especially since you said it could serve other use cases too?

Keinberger commented:
Hey everyone, I made a commit to address the binary split that was discussed above. Now the repository contains two disticnt binaries:
- `miden-faucet-client`: This one holds the existing `miden-client` logic.
- `miden-faucet-operator`: This one contains the mint logic (and any future logic we may want to add). Not sure if the naming is perfect, let me know if you have any other proposals!

Remaining open questions:

- There is a helper function that I implemented here: I borrowed this helper function from the code present at the `miden-client` repository. Since both `miden-client` and `miden-faucet` will now use the same logic, I think it might make sense to include that function in `miden-base` so that both repositories can import it from there. Let me know about your thoughts! If we want to propose changes to this function and decide to move it to `miden-base`, I think we should implement the "changed" function there.
- Since this is being merged onto `next`, will this command only be included at v13? I want this command to be included as soon as possible, would a merge against `main` be the right approach?

Keinberger commented on README.md:
````diff
@@ -25,13 +25,20 @@ miden-faucet init \
 ```
 
+Requesting tokens from a faucet endpoint is handled by the separate `miden-faucet-operator` binary:
+```bash
````
Absolutely agree, I refactored the markdown documentation to adjust for that.

Keinberger commented on bin/faucet-client/src/mint.rs:
```diff
@@ -0,0 +1,438 @@
+                )),
+            })
+    }
+}
```
I agree

bobbinth commented on bin/faucet-operator/Cargo.toml:
```diff
@@ -1,11 +1,11 @@
 edition.workspace      = true
 homepage.workspace     = true
-keywords               = ["faucet", "miden", "node"]
+keywords               = ["faucet", "miden", "node", "operator"]
```
nit: let's remove "node" form the list.

Keinberger replied:
Agreed!

bobbinth commented on bin/faucet-client/Cargo.toml:
```diff
@@ -0,0 +1,33 @@
+homepage.workspace     = true
+license.workspace      = true
+name                   = "miden-faucet-client"
+readme                 = "README.md"
```
It seems like the readme file is missing.

bobbinth replied:
I would probably bring this back and add a short README file (could be very simple - e.g., doesn't need to be more complicated than [this](https://github.com/0xMiden/miden-faucet/tree/next/crates/faucet)).

Keinberger replied:
Agreed, added a simple REDME!

Keinberger commented on Makefile:
```diff
@@ -81,6 +81,7 @@ check: ## Check all targets and features for errors without code generation
 .PHONY: install-faucet
 install-faucet: ## Installs faucet
 	${BUILD_PROTO} cargo install --path bin/faucet --locked
+	${BUILD_PROTO} cargo install --path bin/faucet-operator --locked
```
Sorry, this is a leftover from a previous iteration. In my previous iteration, I kept backwards compatibility for the existing "faucet" binary name. I removed that now, hence why this should also be changed to specify "faucet-client" instead of "faucet".

Thanks for pointing that out!

Keinberger commented on README.md:
````diff
@@ -25,15 +25,23 @@ miden-faucet init \
+Requesting tokens from a faucet endpoint is handled by the separate `miden-faucet-client` binary:
+```bash
+miden-faucet-client mint --url <FAUCET_API_URL> --account <ACCOUNT_ID> --amount <BASE_UNITS>
+```
````
Agreed! I will implement a sub-section

Keinberger commented on docs/src/getting-started/quick-start.md:
```diff
@@ -24,16 +24,31 @@ miden-faucet init \
 ## Step 3: Request Test Tokens
 
-Once the faucet is running, you can request test tokens through either the web interface or the REST API.
+Once the faucet is running, you can request test tokens through either the web interface, the operator CLI, or the REST API.
```
Yes, sorry this is still a leftover from my last refactoring. Will go through the docs files again to make sure that everything is named correctly!

Keinberger commented on docs/src/getting-started/cli.md:
````diff
@@ -210,17 +210,28 @@ Enable OpenTelemetry for production monitoring:
+To see available options:
+```bash
+miden-faucet-operator mint --help
+```
````
Yes, this is indeed wrong. Will fix it!

mmagician commented on bin/faucet-client/src/mint.rs:
```diff
@@ -0,0 +1,309 @@
+// =================================================================================================
+
+const DEFAULT_FAUCET_URL: &str = "https://faucet-api.testnet.miden.io";
+const DEFAULT_TIMEOUT_MS: u64 = 30_000;
```
The "default" part of the name of `DEFAULT_TIMEOUT_MS` makes it sound as if this is configurable - which it is not. The two approaches are:
1. Make it configurable (not worth imo)
2. Rename

Keinberger replied:
Okay, I'll rename it to `REQUEST_TIMEOUT_MS` then, is that fine for you?

Keinberger commented on docs/src/getting-started/cli.md:
```diff
@@ -2,6 +2,11 @@
+The faucet comes with two CLI tools:
+
+- **miden-faucet-operator**: Runs the faucet, used for initializing and starting the faucet.
+- **miden-faucet-client**: Used for interacting with a live faucet, i.e. for requesting tokens from a running faucet.
```
No I made the decision not to have backwards compatibility, as we usually do have breaking changes on different versions.

Nevertheless, since this will be released as part of a patch release, I agree with you that it should be backwards compatible. I will bring back that change.

Keinberger commented on bin/faucet-client/src/mint.rs:
```diff
@@ -0,0 +1,309 @@
+
+/// Response from the `/pow` endpoint.
+#[derive(Debug, Deserialize, Serialize, Clone)]
+pub struct PowResponse {
```
Not quite, the response returns String's and no native faucet types. So we'll still have to convert the strings to our types.

Keinberger replied:
Nevertheless, since the struct is not exported in the `miden-faucet-lib` I'll first have to move it there, export it and then I can reuse it inside of the faucet-client. I will do that.

TomasArrachea commented on bin/faucet-client/src/mint.rs:
```diff
@@ -0,0 +1,297 @@
+    }
+
+    let challenge_bytes = hex::decode(challenge_hex).map_err(MintClientError::InvalidChallenge)?;
+
```
Since we have access to the `miden-pow-rate-limiter` crate, we could parse the challenge from the bytes and use the validation directly from there.

Would be something like:
```rust 
    let challenge = Challenge::try_from(challenge_bytes.as_slice())..map_err(MintClientError::InvalidSerialization)?;

    task::spawn_blocking(move || -> Result<u64, MintClientError> {
        let mut rng = rand::rng();

        loop {
            let nonce: u64 = rng.random();

            if challenge.validate_pow(nonce) {
                return Ok(nonce);
            }
        }
    })
    .await
```

Keinberger commented on bin/faucet-client/src/mint.rs:
```diff
@@ -0,0 +1,297 @@
+            .http_client
+            .get(pow_url)
+            .query(&[("account_id", account_id.to_hex()), ("amount", amount.to_string())]);
+
```
The newly shared types cannot be used directly, as the HTTP request expects raw string types (no native client/faucet types for arguments).

Nevertheless, what we can do is export the raw query parameter structs from the faucet-operator and then use those in the client. I will do that.

Keinberger commented on bin/faucet-client/src/mint.rs:
```diff
@@ -0,0 +1,297 @@
+
+    /// Quantity to mint (in base units).
+    #[arg(short = 'q', long = "quantity", value_name = "U64", alias = "amount")]
+    quantity: u64,
```
Originally I had this named as `amount`, but since `account` and `amount` both would use the `a` short flag, I applied Tomas's suggestion, which he pointed out [here](https://github.com/0xMiden/miden-faucet/pull/196#discussion_r2607084495), and renamed the flag to "quantity".

Given what you have said, I think `amount` and `target` are actually better fitting and more in line with code in our existing Miden repositories.

Keinberger commented on bin/faucet-client/README.md:
```diff
@@ -0,0 +1,13 @@
+
+### `mint`
+
+Requests tokens from a faucet by solving its PoW challenge and receiving a public P2ID note.
```
Agreed, I will change the wording to better reflect what's actually happening here.

Keinberger commented on packaging/faucet/miden-faucet.service:
So in essence you are saying we should keep the file name until the next version?

Keinberger replied:
I refactored the `action.yml` workflow file to make sure that it references all crate, file and directory names correctly, especially this renamed file. In that case, I think the renaming is fine although I do think that someone who is knowledgable about how the deployment works in detail should take a look at my changes to the `action.yml` file.

Keinberger commented on docs/src/getting-started/quick-start.md:
I unfortunately cannot see the lines which you were referring to (maybe my Github UI is broken). Can you create a new comment?

Keinberger commented:
Addressed the comments from @igamigo , as part of that I:
- Renamed the `quantity` and `account` parameters to `amount` and `target-account` respectively
- Implemented the nit changes

There are still some open questions, mainly regarding our CI pipeline:
- I refactored the `action.yml` file to make sure it references the correct directories, file and crate names. Can someone please check if the `action.yml` file is now correct?
- While making changes in the `action.yml` file, I discovered that the `crate_dir` input parameter had a default value of `miden-faucet`, even though the binary folder is currently named `bin/faucet` on `main`. Either I am misunderstanding the purpose of this input parameter or the value was not correct. Would appreciate if someone could check that too!

@TomasArrachea I saw that most of the changes inside of the `action.yml` file came from your commits, would you mind taking a look at my changes of that file to make sure that everything looks correct?

igamigo reviewed (approved):
LGTM! I think @sergerad should probably look into this as it changes some things in relation to how the faucet service was being set up before

sergerad reviewed (approved):
Approved but I _think_ renaming faucet binary to faucet-operator has added work for no value. There will need to be a few small changes in the infrastructure repo following this PR also.

mmagician reviewed (approved):
LGTM with a few leftover naming changes

Keinberger commented:
> @Keinberger, this PR was merged to `main`, is the plan to patch release it? Could you do a PR to `next` bringing these commits (I think this PR was not squash & merged so it is probably multiple)? It seems likely that we'll need to change the target of [0xMiden/miden-client#1642](https://github.com/0xMiden/miden-client/pull/1642) as well

Correct, plan would be a patch release. I will do a PR on `next` to bring those commits onto next.

Sorry about forgetting to squash merge this... Won't happen again!

As for the target of https://github.com/0xMiden/miden-client/pull/1642 - why would we need to change it? If we do a patch release I suppose it also belongs on `main` first, or am I missing something?

Keinberger commented:
FYI I'm also fine with merging everthing onto `next` and introduce the changes as part of v14, in case we are going to release v14 this week.

igamigo commented:
> As for the target of https://github.com/0xMiden/miden-client/pull/1642 - why would we need to change it? If we do a patch release I suppose it also belongs on main first, or am I missing something?

> FYI I'm also fine with merging everthing onto next and introduce the changes as part of v14, in case we are going to release v14 this week.

No need to merge to `main`, we can just merge to `next` and release it on 0.14, and then use it here. Also, the plan is to release 0.14 within today and tomorrow, but it also depends on when the node is released (cc @Dominik1999)

Dominik1999 commented:
It would be the next relase, v0.13, right? 

The plan is to release v0.13 of the client tomorrow or Wednesday. This PR was already merged to `main`, right?
